lib/go-tc/crstates.go

```diff
@@ -25,8 +25,8 @@ import (
 
 // CRStates includes availability data for caches and delivery services, as gathered and aggregated by this Traffic Monitor. It is designed to be served at an API endpoint primarily for Traffic Routers (Content Router) to consume.
 type CRStates struct {
-	Caches          map[CacheName]IsAvailable                       `json:"caches"`
-	DeliveryService map[DeliveryServiceName]CRStatesDeliveryService `json:"deliveryServices"`
+	Caches          map[string]IsAvailable             `json:"caches"`
```

Comment:
How strongly do you feel about all these `string` changes?

So, there were a couple reasons I started using strong typing. 

For one, it self-documents `map[string]string`. I got tired of `//map[CacheName]DeliveryServiceName` everywhere. Worse, if the comments aren't there, it's not always obvious. Even here, `Caches map[string]IsAvailable` could actually be a map[DeliveryServiceName]IsAvailable` for some reason. It's unlikely, but `tc.CacheName` makes it clear for sure.

The other big reason, was that I kept getting them wrong. I'd pass a Cache Group name to something that needed a Cache name. I know that sounds ridiculous and unlikely, but in all the complexity of things like parsing DS stats, I found myself doing it, multiple times. And worse, the failure often isn't obvious. A stat just doesn't get populated, or a lookup thinks a DS doesn't exist, and things _mostly_ work.

That might seem ridiculous, but I really was doing it, repeatedly. Maybe I'm just not as clever as some people, but this kind of thing really helped me at least, to avoid subtle and sometimes very-hard-to-catch errors at IMO a relatively small verbosity cost.

Would you object strenuously if we left these in, and kept using the strong types, to make the code a little more self-documenting and help catch subtle runtime bugs like this?

Reply:
I do really wish that this wasn't done so widely. Really I just wanted to use strings in the code that I was writing, because I found all the casting back and forth obnoxious. But I got so frustrated with recursive compile errors (x is a string now, so you can't pass a cachename as the variable anymore, and then that cachename value from a map can't be assigned to the variable that gets passed into the function etc.) that I threw up my hands and did a search/replace on the whole thing. Which is why `tc.CacheGroupName` is unchanged everywhere - because it wasn't in the code I wrote.

That really does seem pretty ridiculous, but like I said I didn't have to touch anything with Cache Groups, so maybe it wouldn't if I had to deal with more complex things.

I can try to take as much of it out as possible, but I would still like the code I wrote to use strings. Just seems sensible to me. And it really does suck that I had to touch code in `/lib` and `/traffic_ops/traffic_ops_golang`.

Reply:
Alright, I put it back in in most places.